File: src/main/java/org/jukeboxmc/block/behavior/BlockTripwireHook.java
package org.jukeboxmc.block.behavior;

import org.cloudburstmc.nbt.NbtMap;
import org.jukeboxmc.block.Block;
import org.jukeboxmc.block.direction.Direction;
import org.jukeboxmc.util.Identifier;

/**
 * @author LucGamesYT
 * @version 1.0
 */
public class BlockTripwireHook extends Block {

    public BlockTripwireHook( Identifier identifier ) {
        super( identifier );
    }

    public BlockTripwireHook( Identifier identifier, NbtMap blockStates ) {
        super( identifier, blockStates );
    }

    public void setPowered( boolean value ) {
        this.setState( "powered_bit", value ? 1 : 0 );
    }

    public boolean isPowered() {
        return this.stateExists( "powered_bit" ) && this.getByteState( "powered_bit" ) == 1;
    }

    public void setAttached( boolean value ) {
        this.setState( "attached_bit", value ? 1 : 0 );
    }

    public boolean isAttached() {
        return this.stateExists( "attached_bit" ) && this.getByteState( "attached_bit" ) == 1;
    }

    public void setDirection( Direction direction ) {
        switch ( direction ) {
            case SOUTH -> this.setState( "direction", 0 );
            case WEST -> this.setState( "direction", 1 );
            case NORTH -> this.setState( "direction", 2 );
            case EAST -> this.setState( "direction", 3 );
        }
    }

    public Direction getDirection() {
        int value = this.stateExists( "direction" ) ? this.getIntState( "direction" ) : 0;
        return switch ( value ) {
            case 0 -> Direction.SOUTH;
            case 1 -> Direction.WEST;
            case 2 -> Direction.NORTH;
            default -> Direction.EAST;
        };
    }
}

File: JukeboxMC-Server/src/main/kotlin/org/jukeboxmc/server/block/behavior/BlockTripwireHook.kt
package org.jukeboxmc.server.block.behavior

import org.cloudburstmc.nbt.NbtMap
import org.jukeboxmc.api.Identifier
import org.jukeboxmc.api.block.TripwireHook
import org.jukeboxmc.api.block.data.BlockFace
import org.jukeboxmc.api.block.data.Direction
import org.jukeboxmc.api.math.Vector
import org.jukeboxmc.server.block.JukeboxBlock
import org.jukeboxmc.server.extensions.isOneOf
import org.jukeboxmc.server.extensions.toByte
import org.jukeboxmc.server.item.JukeboxItem
import org.jukeboxmc.server.player.JukeboxPlayer
import org.jukeboxmc.server.world.JukeboxWorld

class BlockTripwireHook(identifier: Identifier, blockStates: NbtMap?) : JukeboxBlock(identifier, blockStates),
    TripwireHook {

    override fun placeBlock(
        player: JukeboxPlayer,
        world: JukeboxWorld,
        blockPosition: Vector,
        placePosition: Vector,
        clickedPosition: Vector,
        itemInHand: JukeboxItem,
        blockFace: BlockFace
    ): Boolean {
        if (blockFace.isOneOf(BlockFace.UP, BlockFace.DOWN)) return false
        this.setDirection(blockFace.toDirection())
        return super.placeBlock(player, world, blockPosition, placePosition, clickedPosition, itemInHand, blockFace)
    }

    override fun getDirection(): Direction {
        return when (this.getIntState("direction")) {
            0 -> Direction.SOUTH
            1 -> Direction.WEST
            2 -> Direction.NORTH
            else -> Direction.EAST
        }
    }

    override fun setDirection(value: Direction): TripwireHook {
        return when (value) {
            Direction.SOUTH -> this.setState("direction", 0)
            Direction.WEST -> this.setState("direction", 1)
            Direction.NORTH -> this.setState("direction", 2)
            else -> this.setState("direction", 3)
        }
    }

    override fun isAttached(): Boolean {
        return this.getBooleanState("attached_bit")
    }

    override fun setAttached(value: Boolean): TripwireHook {
        return this.setState("attached_bit", value.toByte())
    }

    override fun isPowered(): Boolean {
        return this.getBooleanState("powered_bit")
    }

    override fun setPowered(value: Boolean): TripwireHook {
        return this.setState("powered_bit", value.toByte())
    }
}